File: Python/0547-m-friend-circles.py
"""
DFS
Time: O(n^2)
Space: O(n)
"""
class Solution_1:
    def findCircleNum(self, M: List[List[int]]) -> int:
        def dfs(M, curr, n):
            for i in range(n):
                if M[curr][i] == 1:
                    M[curr][i] = M[i][curr] = 0
                    dfs(M, i, n)

        n = len(M)
        res = 0
        for i in range(n):
            if M[i][i] == 1:
                res += 1
                dfs(M, i, n)

        return res

"""
Union Find
"""
class Solution_2:
    def findCircleNum(self, M: List[List[int]]) -> int:
        n = len(M)
        parent = [-1] * n
        for i in range(n):
            for j in range(n):
                if M[i][j] == 1 and i != j:
                    self.union(parent, i, j)
        res = 0
        for i in range(n):
            if parent[i] == -1:
                res += 1
        return res

    def find_parent(self, parent, i):
        path = []
        while parent[i] != -1:
            path.append(i)
            i = parent[i]

        for p in path:
            parent[p] = i

        return i

    def union(self, parent, x, y):
        x_set = self.find_parent(parent, x)
        y_set = self.find_parent(parent, y)
        if x_set != y_set:
            parent[x_set] = y_set

File: leetcode/C++/0547-m-friend-circles.cpp
// DFS
class Solution {
private:
    void dfs(vector<vector<int>>& M, int cur, int n) {
        for (int i = 0; i < n; i++) {
            if (M[cur][i] == 1) {
                M[cur][i] = 0;
                M[i][cur] = 0;
                dfs(M, i, n);
            }
        }
    }
    
public:
    int findCircleNum(vector<vector<int>>& M) {
        int n = M.size();
        int res = 0;
        for (int i = 0; i < n; i++) {
            if (M[i][i] == 1) {
                res++;
                dfs(M, i, n);
            }
        }
        return res;
    }
};

// Union Find
class Solution {
public:
    int findCircleNum(vector<vector<int>>& M) {
        int n = M.size();
        vector<int> parent(n, -1);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (M[i][j] == 1 && i != j) {
                    union_op(parent, i, j);
                }
            }
        }
        int res = 0;
        for (int i = 0; i < n; i++) {
            if (parent[i] == -1) res++;
        }
        return res;
    }
    
    int find_parent(vector<int>& parent, int i) {
        vector<int> path;
        while (parent[i] != -1) {
            path.push_back(i);
            i = parent[i];
        }
        
        for (int p : path)
            parent[p] = i;
        return i;
    }
    
    void union_op(vector<int>& parent, int x, int y) {
        int x_set = find_parent(parent, x);
        int y_set = find_parent(parent, y);
        if (x_set != y_set)
            parent[x_set] = y_set;
    }
};